### ISSUE TITLE
Export problems

### ISSUE THREAD

**Issue by axomat** [octobercms/october#1313]
Hi
Its great to see the new options in Import and Export but here is some feedback.

1. I had to remove the debugbar plugin which adds html to the exported file and causes the browser to think it is an html page. Removing the plugin makes it work but the browser still thinks the csv file is an html page so I think there is a download type issue somewhere.

2. I cant find a way to get eloquent to get() an attribute appended to the model. It only works if you get all columns which may be a limitation of laravel but annoying if, like me, you make a lot of use of appended attributes. This worked in the old method.
EDIT:
I have solved this by using the following in the import/export model
return Subscriber::get()->toArray();
instead of 
return Subscriber::get($columns);
which is how the documentation suggests it should be used.

3. Exporting columns that have been created from relations is much more difficult (it was easy in the old method), in fact I dont know how to do it but I presume some laravel magic in the import/export model would do it,

So in conclusion, please could we have the old "simple" export back as an alternative to the new one? It worked really easily, the user could do a search and refine the list of records before downloading them (this does not work at all with the new method), and it was really easy to understand for the user "if you can see the data on the page, you will see it in the downloaded file". As a programmer it was really easy as well, just a line in a config file. I used it in nearly all my controllers, instead now I have to go through all my plugins and rewrite to avoid the issues above.

Please?!

**Comment by axomat** [comment]
A further update on this, I have reproduced a search option by using the form facility. For anyone else trying this, I had to make the fields used in the form all fillable in the importexport model.

**Comment by daftspunk** [comment]
I was going to suggest toArray(), perhaps I will add this to the documentation. The debugbar needs to detect if the response is not text/html and disable itself, this will cause problems for the Theme Export feature too. Exporting with relations is straight forward, simply make the columns visible and toArray() will take care of the rest. Eg:

```
protected $visible = ['somerelation.somefield'];
```

Or see the `addVisible()` method.

**Comment by axomat** [comment]
Thanks Daftspunk fantastic response as always!

**Comment by Floriskoch** [comment]
Where can I find some more documentation on this? I'm trying to export own plugin but the relation fields stay empty. I've looked here https://octobercms.com/docs/backend/import-export but it doesn't say anything about exporting relations. The above comments also don't work. I've swapped the `protected $visible` property for a `public $attributes` property but that does not work either. Would be great to get some more guidance on this subject. Thanks!

**Comment by chrisvidal** [comment]
Same here, I can't seem to find any documentation on how to export simply a relation or 2 levels of relation also

**Comment by roojai** [comment]
Me Too. Its working in the blog plugin ok though, just can't seem to copy that code and get it to work properly!

**Comment by roojai** [comment]
Only 12 mins after the post I get it working haha (after the whole day on this).

**Comment by LukeTowers** [comment]
Mind sharing the stumbling blocks and your end solution? @roojai

**Comment by roojai** [comment]
@LukeTowers  sure no problem. In my case I was exporting/importing products that had many to many relationship with two other models - brands and categories.

The working code for the export model was as follows:

`<?php namespace Roojai\Eutsproducts\Models;

use Backend\Models\ExportModel;
use ApplicationException;

/**
 * Post Export Model
 */
class ProductExport extends ExportModel
{
    public $table = 'roojai_eutsproducts_products';

    /**
     * @var array Relations
     */

    'product_brands' => [
      		'Roojai\EutsProducts\Models\Brand',
      		'table' => 'roojai_eutsproducts_products_brands',
          'key' => 'product_id',
          'otherKey' => 'brand_id'
      	],


      'product_categories' => [
      		'Roojai\EutsProducts\Models\Category',
      		'table' => 'roojai_eutsproducts_products_categories',
          'key' => 'product_id',
          'otherKey' => 'category_id'
      	]
      	];

        /**
         * The accessors to append to the model's array form.
         * @var array
         */
        protected $appends = [
            'categories',
            'brands',
        ];

    public function exportData($columns, $sessionKey = null)
    {
      $result = self::make()
      ->with([
        'product_brands',
        'product_collections'
      ])
      ->get()
      ->toArray()
  ;

  return $result;
}
public function getCategoriesAttribute()
{
    if (!$this->product_categories) {
        return '';
    }

    return $this->encodeArrayValue($this->product_categories->lists('name'));
}
public function getBrandsAttribute()
{
    if (!$this->product_brands) {
        return '';
    }

    return $this->encodeArrayValue($this->product_brands->lists('name'));
}

}`
